tket/src/passes/squash_borrow.rs

```diff
@@ -0,0 +1,708 @@
+    }
+    Ok(elided)
+}
+
+fn next_array_op(
+    hugr: &impl HugrView<Node = Node>,
+    is_br: &impl IsBorrowReturn,
+    candidates: &mut impl Extend<Wire>,
+    array: Wire,
+    array_ty: &Type,
+) -> Result<Option<(Node, u64, BRAction, BorrowFromPorts)>, BorrowAnalysisError> {
+    let (node, inport) = hugr
+        .single_linked_input(array.node(), array.source())
+        .expect("array is linear");
+    let Some(is_br) = is_br.is_borrow_return(node, hugr)? else {
+        // Not a borrow/return - stop processing this array;
+        // any outports of the node are considered fresh arrays.
+        candidates.extend(all_outs(hugr, node));
+        return Ok(None);
+    };
+
+    if inport != is_br.borrow_from.inc {
+        if let BRAction::Borrow(_) = is_br.action {
+            panic!("Array fed into unexpected port of borrow")
+        }
+        // Op will be reachable along the incoming-array port, so some other call
+        // (along that path) will process it.
+        if let BRAction::Return(rv) = is_br.action {
+            assert_eq!(rv, inport); // Nested array being returned into outer array.
+        }
+        return Ok(None);
+    }
+
+    {
+        let out_ty = wire_type(hugr, Wire::new(node, is_br.borrow_from.out));
+        if *array_ty != out_ty {
+            let array_ty = array_ty.clone();
+            return Err(BorrowAnalysisError::InconsistentArrayType { array_ty, out_ty });
+        }
+    }
+    let Some(idx) = find_const(hugr, node, is_br.elem_index) else {
+        // Unknown index.
+        // Hence, we must preserve borrowedness of all indices for this op;
+        // so we can't elide anything before this op with anything after.
+        // Hence, op can be considered as beginning of fresh array(s) on each outport.
+        candidates.extend(all_outs(hugr, node));
+        return Ok(None);
+    };
+    if matches!(is_br.action, BRAction::Clobber(_)) {
+        // We'll process the borrow-from-out port here, but the others are potentially fresh arrays.
+        candidates.extend(all_outs(hugr, node).filter(|w| w.source() != is_br.borrow_from.out));
+    }
+    Ok(Some((node, idx, is_br.action, is_br.borrow_from)))
+}
+
+fn wire_type(h: &impl HugrView<Node = Node>, w: Wire) -> Type {
+    h.out_value_types(w.node())
+        .find(|(p, _)| *p == w.source())
+        .unwrap()
+        .1
+}
+
+fn all_outs(h: &impl HugrView<Node = Node>, n: Node) -> impl Iterator<Item = Wire> + '_ {
+    h.out_value_types(n).map(move |(p, _)| Wire::new(n, p))
+}
+
+fn find_const<H: HugrView>(hugr: &H, n: H::Node, inp: IncomingPort) -> Option<u64> {
+    let (mut curr_node, mut outp) = hugr.single_linked_output(n, inp).unwrap();
+
+    fn is_const_conversion_op(op: &OpType) -> bool {
+        matches!(op, OpType::LoadConstant(..))
+            || op
+                .as_extension_op()
+                .is_some_and(|op| ConvertOpDef::from_extension_op(op) == Ok(ConvertOpDef::itousize))
```

Comment:
this feels flaky - future conversion ops wouldn't be covered. Would it not be better to rely on constant folding eliminating these beforehand?